tfhe/src/integer/gpu/ciphertext/compressed_ciphertext_list.rs

```diff
@@ -0,0 +1,205 @@
+            end_block_index,
+            streams,
+            gpu_index,
+        )
+    }
+}
+
+pub trait CudaCompressible {
+    fn compress_into(
+        self,
+        messages: &mut Vec<CudaRadixCiphertext>,
+        streams: &CudaStreams,
+    ) -> DataKind;
+}
+
+// Todo: Can we combine these two impl using CudaIntegerRadixCiphertext?
+impl CudaCompressible for CudaSignedRadixCiphertext {
+    fn compress_into(
+        self,
+        messages: &mut Vec<CudaRadixCiphertext>,
+        streams: &CudaStreams,
+    ) -> DataKind {
+        let x = self.ciphertext.duplicate(streams);
+
+        let copy = x.duplicate(streams);
+        messages.push(copy);
+
+        let num_blocks = x.d_blocks.lwe_ciphertext_count().0;
+        DataKind::Signed(num_blocks)
+    }
+}
+impl CudaCompressible for CudaUnsignedRadixCiphertext {
+    fn compress_into(
+        self,
+        messages: &mut Vec<CudaRadixCiphertext>,
+        streams: &CudaStreams,
+    ) -> DataKind {
+        let x = self.ciphertext.duplicate(streams);
+
+        let copy = x.duplicate(streams);
+        messages.push(copy);
+
+        let num_blocks = x.d_blocks.lwe_ciphertext_count().0;
+
+        DataKind::Unsigned(num_blocks)
+    }
+}
+
+pub struct CudaCompressedCiphertextListBuilder {
+    pub(crate) ciphertexts: Vec<CudaRadixCiphertext>,
+    pub(crate) info: Vec<DataKind>,
+}
+
+impl CudaCompressedCiphertextListBuilder {
+    #[allow(clippy::new_without_default)]
+    pub fn new() -> Self {
+        Self {
+            ciphertexts: vec![],
+            info: vec![],
+        }
+    }
+
+    pub fn push<T: CudaCompressible>(&mut self, data: T, streams: &CudaStreams) -> &mut Self {
+        let kind = data.compress_into(&mut self.ciphertexts, streams);
+
+        if kind.num_blocks() != 0 {
+            self.info.push(kind);
+        }
+
+        self
+    }
+
+    pub fn build(
+        &self,
+        comp_key: &CudaCompressionKey,
+        streams: &CudaStreams,
+        gpu_index: u32,
+    ) -> CudaCompressedCiphertextList {
+        let packed_list =
+            comp_key.compress_ciphertexts_into_list(&self.ciphertexts, streams, gpu_index);
+
+        CudaCompressedCiphertextList {
+            packed_list,
+            info: self.info.clone(),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::integer::gpu::gen_keys_radix_gpu;
+    use crate::shortint::parameters::list_compression::COMP_PARAM_MESSAGE_2_CARRY_2_KS_PBS_TUNIFORM_2M64;
+    use crate::shortint::prelude::PARAM_MESSAGE_2_CARRY_2_KS_PBS;
+
+    #[test]
+    fn test_gpu_ciphertext_compression() {
+        let gpu_index = 0;
+        let streams = CudaStreams::new_single_gpu(gpu_index);
+
+        let num_blocks = 4;
+        let (cks, _) = gen_keys_radix_gpu(PARAM_MESSAGE_2_CARRY_2_KS_PBS, num_blocks, &streams);
+
+        let private_compression_key =
+            cks.new_compression_private_key(COMP_PARAM_MESSAGE_2_CARRY_2_KS_PBS_TUNIFORM_2M64);
```

Comment:
Can we really use these parameters? Shouldn't it be a gaussian noise?

Reply:
I don't know. `TUNIFORM` works but so do `GAUSSIAN` parameters.
I will replace the ones used here by Gaussian since its what we are used to work with in the GPU.